rustdoc: add hash to filename of toolchain files

jsha opened the pull request:
All static files used by rustdoc are now stored in static.files/ and their filenames include a hash of their contents. Their filenames no longer include the contents of the --resource-suffix flag. This clarifies caching semantics. Anything in static.files can use Cache-Control: immutable because any updates will show up as a new URL.

Invocation-specific files like crates-NN.js, search-index-NN.js, and sidebar-items-NN.js still get the resource suffix.

This has a useful side effect: once toolchain files aren't affected by resource suffix, it will become possible for docs.rs to include crate version in the resource suffix. That should fix a caching issue with `/latest/` URLs: https://github.com/rust-lang/docs.rs/issues/1593. My goal is that it should be safe to serve all rustdoc JS, CSS, and fonts with infinite caching headers, even when new versions of a crate are uploaded in the same place as old versions.

The --disable-minification flag is removed because it would vary the output of static files based on invocation flags. Instead, for rustdoc development purposes it's preferable to symlink static files to a non-minified copy for quick iteration.

Example listing:

```
$ cd build/x86_64-unknown-linux-gnu/doc/ && find . | egrep 'js$|css$' | egrep -v 'sidebar-items|implementors' | sort
./crates1.65.0.js
./rust.css
./search-index1.65.0.js
./source-files1.65.0.js
./static.files/ayu-2bfd0af01c176fd5.css
./static.files/dark-95d11b5416841799.css
./static.files/light-c83a97e93a11f15a.css
./static.files/main-efc63f77fb116394.js
./static.files/normalize-76eba96aa4d2e634.css
./static.files/noscript-5bf457055038775c.css
./static.files/rustdoc-7a422337900fa894.css
./static.files/scrape-examples-3dd10048bcead3a4.js
./static.files/search-47f3c289722672cf.js
./static.files/settings-17b08337296ac774.js
./static.files/settings-3f95eacb845293c0.css
./static.files/source-script-215e9db86679192e.js
./static.files/storage-26d846fcae82ff09.js
```

Fixes #98413

rust-highfive commented:
r? @GuillaumeGomez

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8295210294?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/a89d02970a738a3f02fba8b63320802044e2db66/checks/8295210294/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
info: ES-Check: there were no ES version matching errors!  🎉
+ eslint -c ../src/librustdoc/html/static/.eslintrc.js ../src/librustdoc/html/static/js/externs.js ../src/librustdoc/html/static/js/main.js ../src/librustdoc/html/static/js/scrape-examples.js ../src/librustdoc/html/static/js/search.js ../src/librustdoc/html/static/js/settings.js ../src/librustdoc/html/static/js/source-script.js ../src/librustdoc/html/static/js/storage.js

/checkout/src/librustdoc/html/static/js/settings.js
  150:64  error  Missing space before =>  arrow-spacing
  150:67  error  Missing space after =>   arrow-spacing
/checkout/src/librustdoc/html/static/js/storage.js
/checkout/src/librustdoc/html/static/js/storage.js
  151:28  error  Expected '!==' and instead saw '!='  eqeqeq

✖ 3 problems (3 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.
```

</details>

Nemo157 commented:
It will be necessary to figure out how docs.rs will handle this. Currently the shared resource handling is a massive hack that allows a static resource request from anywhere but will only lookup the file from `/` (so won't work if those files are in a sub-directory). This is probably a good opportunity to figure out how to improve that for future doc builds.

<https://github.com/rust-lang/docs.rs/blob/4abe7e9c9bd1298ec3af01878af554a5b60797e7/src/web/mod.rs#L165-L177>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/8296891402?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/3d75fc8c4d3671e0ca7a6192091d06a9e8e04645/checks/8296891402/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Rustbook (x86_64-unknown-linux-gnu) - embedded-book
Rustbook (x86_64-unknown-linux-gnu) - edition-guide
Building stage0 tool linkchecker (x86_64-unknown-linux-gnu)
    Finished release [optimized] target(s) in 0.13s
guide-tasks.html:9: broken link - `rustdoc1.65.0.css`
guide-macros.html:9: broken link - `rustdoc1.65.0.css`
tutorial.html:9: broken link - `rustdoc1.65.0.css`
guide-crates.html:9: broken link - `rustdoc1.65.0.css`
guide-pointers.html:9: broken link - `rustdoc1.65.0.css`
complement-lang-faq.html:9: broken link - `rustdoc1.65.0.css`
guide-ffi.html:9: broken link - `rustdoc1.65.0.css`
guide-strings.html:9: broken link - `rustdoc1.65.0.css`
grammar.html:9: broken link - `rustdoc1.65.0.css`
rustdoc.html:9: broken link - `rustdoc1.65.0.css`
guide.html:9: broken link - `rustdoc1.65.0.css`
reference.html:9: broken link - `rustdoc1.65.0.css`
complement-project-faq.html:9: broken link - `rustdoc1.65.0.css`
guide-testing.html:9: broken link - `rustdoc1.65.0.css`
complement-design-faq.html:9: broken link - `rustdoc1.65.0.css`
guide-error-handling.html:9: broken link - `rustdoc1.65.0.css`
guide-unsafe.html:9: broken link - `rustdoc1.65.0.css`
rust.html:9: broken link - `rustdoc1.65.0.css`
guide-plugins.html:9: broken link - `rustdoc1.65.0.css`
guide-ownership.html:9: broken link - `rustdoc1.65.0.css`
intro.html:9: broken link - `rustdoc1.65.0.css`
number of HTML files scanned: 32979
number of HTML redirects found: 10114
number of links checked: 2473159
number of links ignored due to external: 118066
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #101895) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3085961803/jobs/4989833527) [(plain)](https://github.com/rust-lang/rust/commit/749fa1a80bcfb08586a90b84d8aa0708ecd89e57/checks/8438433384/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
doc tests for: /checkout/src/doc/rustc/src/tests/index.md
doc tests for: /checkout/src/doc/rustc/src/what-is-rustc.md
 finished in 1.491 seconds
Generating lint docs (x86_64-unknown-linux-gnu)
internal/modules/cjs/loader.js:818
  throw err;

Error: Cannot find module '/checkout/obj/build/x86_64-unknown-linux-gnu/doc/search1.66.0.js'
Require stack:
- /checkout/src/tools/rustdoc-js/tester.js
- /checkout/src/tools/rustdoc-js/tester.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:815:15)
    at Function.Module._load (internal/modules/cjs/loader.js:667:27)
    at Module.require (internal/modules/cjs/loader.js:887:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at loadSearchJS (/checkout/src/tools/rustdoc-js/tester.js:313:26)
    at main (/checkout/src/tools/rustdoc-js/tester.js:391:26)
    at Object.<anonymous> (/checkout/src/tools/rustdoc-js/tester.js:418:14)
    at Module._compile (internal/modules/cjs/loader.js:999:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1027:10)
    at Module.load (internal/modules/cjs/loader.js:863:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/checkout/src/tools/rustdoc-js/tester.js' ]
Build completed unsuccessfully in 0:26:42
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3086315189/jobs/4990542812) [(plain)](https://github.com/rust-lang/rust/commit/0bf7ad318571fcbbd5db9387c2db066bd7895131/checks/8439392763/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

Check compiletest suite=run-make mode=run-make (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 67 tests
Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
iii...i.i.F....iiiii..............ii.i..iiiiiii.iiii........F......

---- [run-make] src/test/run-make/issue-88756-default-output stdout ----

error: make failed
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/issue-88756-default-output/issue-88756-default-output:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc' 2>&1 | sed -E 's@/nightly/|/beta/|/stable/|/1\.[0-9]+\.[0-9]+/@/$CHANNEL/@g' | diff - output-default.stdout
>         --disable-minification 
>         --disable-minification 
>                         Disable minification applied on JS files
<         --disable-minification 
<                         removed
------------------------------------------
--- stderr -------------------------------
---
error: make failed
status: exit status: 2
command: "make"
--- stdout -------------------------------
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc' -L /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib -Z unstable-options --emit=invocation-specific --output /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only --resource-suffix=-xxx --theme y.css --extend-css z.css x.rs
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/search-index-xxx.js ]
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/settings.html ]
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/x/all.html ]
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/x/index.html ]
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/theme-xxx.css ] # generated from z.css
! [ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/storage-xxx.js ]
! [ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/SourceSerif4-It.ttf.woff2 ]
# FIXME: this probably shouldn't have a suffix
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/y-xxx.css ]
# FIXME: this is technically incorrect (see `write_shared`)
! [ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/invocation-only/main-xxx.js ]
LD_LIBRARY_PATH="/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files:/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-bootstrap-tools/x86_64-unknown-linux-gnu/release/deps:/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/lib" '/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc' -L /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib -Z unstable-options --emit=toolchain-shared-resources --output /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/toolchain-only --resource-suffix=-xxx --extend-css z.css x.rs
[ -e /checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/emit-shared-files/emit-shared-files/toolchain-only/storage-xxx.js ]
--- stderr -------------------------------
--- stderr -------------------------------
warning: theme file "y.css" is missing CSS rules from the default theme
  = warning: the theme may appear incorrect when loaded
  = warning: the theme may appear incorrect when loaded
  = help: to see what rules are missing, call `rustdoc --check-theme "y.css"`

make: *** [Makefile:26: toolchain-only] Error 1



failures:
```

</details>

jsha commented:
> It will be necessary to figure out how docs.rs will handle this. Currently the shared resource handling is a massive hack that allows a static resource request from anywhere but will only lookup the file from / (so won't work if those files are in a sub-directory). This is probably a good opportunity to figure out how to improve that for future doc builds.
>
> https://github.com/rust-lang/docs.rs/blob/4abe7e9c9bd1298ec3af01878af554a5b60797e7/src/web/mod.rs#L165-L177

I filed https://github.com/rust-lang/docs.rs/pull/1863 documenting my current understanding of how this hack works.

Assuming that's correct, I think there's a fairly straightforward way to make docs.rs' SharedResourceHandler work with this change: We can add code such that if the path begins with `/static.files/`, SharedResourceHandler looks up the _full_ path in storage, rather than just the last component of the path. So long as that change rolls out to docs.rs before this PR lands, we should be good.

jsha committed:
Make --static-root-path point to static.files

jsha commented:
Another possible way to interface with docs.rs is demonstrated here: https://github.com/rust-lang/rust/commit/ea51f8a13d20af173716811ce0b5943ebfd5b997

In other words, the default value of `--static-root-path` is `../static.files/` (insert appropriate number of `../` depending on depth). The nice thing about this approach is that it makes it much harder to miss a mistake like the one fixed in https://github.com/rust-lang/rust/pull/83094 (confusing "root path" with "static root path" because they default to the same value).

If we adopt this approach, docs.rs would have to make some changes in sync with the rustdoc release:

 - pass `--static-root-path /rustdoc.static/` instead of `--static-root-path /`.
 - for each new toolchain, after running with `--emit=toolchain-shared-resources`, copy the contents of `doc/static.files/` to `/rustdoc.static/` in storage.
 
 (note I specifically chose the different directory name `rustdoc.static` to make it clear it doesn't have to have the same name).

Nemo157 commented:
I like being able to configure the static-root-path explicitly. To make synchronization easier I wonder if we could land that flag first (with it being a no-op), then update docs.rs to use it and copy `doc/static.files` if it exists, then land this change using the flag.

jsha commented:
I'm a little confused - `--static-root-path` already exists, and docs.rs sets it to `/`. So, no need to land a flag.

That said, we _could_ make a change to docs.rs ahead of this one:

 - start storing essential files in /rustdoc.static/
 - start passing `--static-root-path /rustdoc.static/`
 - adjust SharedResourceHandler to treat any paths beginning with `/rustdoc.static/` as full paths rather than just filenames.

Nemo157 commented:
Ah, for some reason I thought the current flag was `--root-path`. So IIUC with this change nothing would actually _need_ to change on the docs.rs side (we would just start getting files with the hash in the "essential files" instead of the resource-suffix). But it would probably still be good for us to change the path to make it easier to apply caching headers and make it more understandable.

With https://github.com/rust-lang/rust/commit/ea51f8a13d20af173716811ce0b5943ebfd5b997 we _would_ have to change something because `--static-root-path` would correspond to `docs/static.files/*` rather than `docs/* (we could pre-prepare for that by detecting whether the `static.files` directory exists and either copy from in there or from the root).

GuillaumeGomez commented on src/librustdoc/config.rs:
```diff
@@ -237,9 +237,6 @@ pub(crate) struct RenderOptions {
     pub(crate) resource_suffix: String,
-    /// Whether to run the static CSS/JavaScript through a minifier when outputting them. `true` by
-    /// default.
-    pub(crate) enable_minification: bool,
```
Since it's a stable option, I don't think it can be removed just like that. We need a FCP or at least to deprecate it first. It was very useful for testing rustdoc too but I'm not sure if anyone other than me is using this option, so I \*think\* it's fine to remove it.

GuillaumeGomez replied:
Then problem solved. :D

GuillaumeGomez commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -4,15 +4,15 @@
 	font-weight: 400;
 	src: local('Fira Sans'),
-		url("FiraSans-Regular.woff2") format("woff2");
+		url("FiraSans-Regular-018c141bf0843ffd.woff2") format("woff2");
```
Does it mean we will have to manually update this hash every time we update these files?

jsha replied:
Yes. Fortunately, the only files referenced in this way in rustdoc.css almost never change: the fonts and the toggles. As a bonus, we get rid of the awkward AUTOREPLACE functionality.

GuillaumeGomez replied:
I'm not sure to be a big fan of this considering it means more manual code to handle. The worst part being that the file will very likely exist in the destination folder already, so if we forget to update this entry, it'll still pick the old one. Remembering to change this will be a nightmare...

jsha replied:
The previous revision of a changed file would only be present in the destination folder for an unclean build (e.g. when developing locally). In CI it would not be present. So, how can we make sure we detect such cases in CI? Is it possible to change browser-ui-test so it treats any failure to load a subresource as failing the test?

GuillaumeGomez replied:
Currently it only detects JS errors. I think we can make it detect load errors too.

GuillaumeGomez commented:
I think there are a few issues coming up from this change. If I understand correctly, the `--resource-suffix` option will be added to the filename in addition to the file's hash. For predictability, I'm not sure if it's a good thing (or if we even care).

The content of `static_files.rs` is duplicated between the array and the struct, which isn't great. I don't think it can be done much better than that though. Maybe we could make a macro I guess so everything would be set in one place to prevent a mismatch between the struct fields and the array elements. What do you think?

I'm really not a big fan to have to update `css` files manually when a static file is updated. The `AUTOREPLACE` thing wasn't great but at least it was automated. It should be pretty rare though since we don't update them often so I guess it's a minor concern.

Another very minor concern: why `static.files` as folder name and not `static_files` for example?

One last point: I think we'll need a perf check to see if there is any impact with having such a big struct (I think not but better be aware of it if there is an impact).

Overall it seems like a really nice improvement. :)

jsha commented:
> If I understand correctly, the --resource-suffix option will be added to the filename in addition to the file's hash. For predictability, I'm not sure if it's a good thing (or if we even care).

No, there are two types of files now:

 - static files (they get a hash)
 - invocation specific files (they get a `--resource-suffix`)
 
 > Another very minor concern: why static.files as folder name and not static_files for example?

I wanted a name that couldn't conflict with a package or module name. I don't feel strongly about the choice and am happy to take suggestions.

> Maybe we could make a macro I guess so everything would be set in one place to prevent a mismatch between the struct fields and the array elements. What do you think?

This is a good idea. I'll see what I can do.

bors commented:
:umbrella: The latest upstream changes (presumably #102726) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

GuillaumeGomez commented:
I wrote this macro:

```rust
pub(crate) struct StaticFile {
    pub(crate) filename: &'static str,
    pub(crate) bytes: &'static [u8],
}

macro_rules! static_files {
    ($($field:ident => $file_path:literal,)+) => {
        pub(crate) struct StaticFiles {
            $($field: StaticFile,)+
        }
        
        pub(crate) const STATIC_FILES: StaticFiles = StaticFiles {
            $($field: StaticFile { filename: $file_path, bytes: include_bytes!($file_path) },)+
        };
        
        pub(crate) static STATIC_FILES_LIST: &[&'static StaticFile] = &[
            $(&STATIC_FILES.$field,)+
        ];
    }
}

// Can be used like this:
static_files!(
    rustdoc_css => "rustdoc.css",
    whatever_css => "whatever.css",
);
```

You can test it online with a little modified version (without the `include_bytes!` part) [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=78aa019eecf1e355a469757149feea45).

jsha commented:
Thanks for writing up that macro! I've updated the branch to use it.

Note that this will need a small adjustment from docs.rs before it lands: https://github.com/rust-lang/docs.rs/pull/1889. But it is ready for another round of review now.

GuillaumeGomez commented:
Thanks again for working on this! r=me once docs.rs is ready.

bors commented:
:umbrella: The latest upstream changes (presumably #103344) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

jsha committed:
rustdoc: add hash to filename of toolchain files

All static files used by rustdoc are now stored in static.files/ and
include a hash of their contents. They no longer include the contents of
the --resource-suffix flag. This clarifies caching semantics. Anything
in static.files can use Cache-Control: immutable because any updates
will show up as a new URL.

Invocation-specific files like crates-NN.js, search-index-NN.js,
and sidebar-items-NN.js still get the resource suffix.

The --disable-minification flag is removed because it would vary the
output of static files based on invocation flags. Instead, for
rustdoc development purposes it's preferable to symlink static files
to a non-minified copy for quick iteration.

jsha commented:
@bors try

Generating artifacts to be able to test rustwide building with this change for docs.rs.

bors commented:
:hourglass: Trying commit 9819574fdc8fe7740f78fb36c60b724acca7cd53 with merge be6926ce9497650b6502108ba836f0f4bdd9b1a4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9819574fdc8fe7740f78fb36c60b724acca7cd53","merge_sha":"be6926ce9497650b6502108ba836f0f4bdd9b1a4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3311181016/jobs/5468592075)
Build commit: be6926ce9497650b6502108ba836f0f4bdd9b1a4 (`be6926ce9497650b6502108ba836f0f4bdd9b1a4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3311181016/jobs/5468592075"},"merge_sha":"be6926ce9497650b6502108ba836f0f4bdd9b1a4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3311181016/jobs/5468592075)
Build commit: be6926ce9497650b6502108ba836f0f4bdd9b1a4 (`be6926ce9497650b6502108ba836f0f4bdd9b1a4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3311181016/jobs/5468592075"},"merge_sha":"be6926ce9497650b6502108ba836f0f4bdd9b1a4"} -->

jsha commented:
Earlier in this thread we talked about a tweak to the semantics of `--static-root-path`, so it points to the directory where the contents of `static.files/` will be found: https://github.com/rust-lang/rust/commit/ea51f8a13d20af173716811ce0b5943ebfd5b997.

In my head I resolved that conversation in favor of incorporating the tweak, and merged it into this branch. However, in actuality I never actually did. So I've cherry-picked it now. The matching PR for docs.rs to use the new `--static-root-path` semantics is https://github.com/rust-lang/docs.rs/pull/1889, which should merge first.

@GuillaumeGomez I'd appreciate another round of review given the new commit.

Also:

@bors try

bors commented:
:hourglass: Trying commit 89388fca949ce79b344f089e45fa5bf32e46f7ba with merge d9e9ccf3bd7f7fab8581081a2462ec47f602103b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"89388fca949ce79b344f089e45fa5bf32e46f7ba","merge_sha":"d9e9ccf3bd7f7fab8581081a2462ec47f602103b"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3326365058/jobs/5500014912) [(plain)](https://github.com/rust-lang/rust/commit/89388fca949ce79b344f089e45fa5bf32e46f7ba/checks/9109344303/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [rustdoc] src/test/rustdoc/static-root-path.rs stdout ----

error: htmldocck failed!
status: exit status: 1
command: "/usr/bin/python3" "/checkout/src/etc/htmldocck.py" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc/static-root-path" "/checkout/src/test/rustdoc/static-root-path.rs"
stdout: none
--- stderr -------------------------------
4: @matchesraw check failed
 `PATTERN` did not match
 // @matchesraw - '"/cache/static.files/main-'
11: @matchesraw check failed
 `PATTERN` did not match
 // @matchesraw - '"/cache/static.files/source-script-'
17: @matchesraw check failed
 `PATTERN` did not match
 // @matchesraw - '/cache/static.files/settings-'
Encountered 3 errors
------------------------------------------


```

</details>

GuillaumeGomez commented:
Let's run a perf check before approving it just in case.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit ae4c3c91e25a5ffbee4c54d0fabc538fd48930b9 with merge 62fd07b1ee7441dad3ff012468c1251d889a5f5b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"ae4c3c91e25a5ffbee4c54d0fabc538fd48930b9","merge_sha":"62fd07b1ee7441dad3ff012468c1251d889a5f5b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3330169062/jobs/5510338921)
Build commit: 62fd07b1ee7441dad3ff012468c1251d889a5f5b (`62fd07b1ee7441dad3ff012468c1251d889a5f5b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3330169062/jobs/5510338921"},"merge_sha":"62fd07b1ee7441dad3ff012468c1251d889a5f5b"} -->

rust-timer commented:
Queued 62fd07b1ee7441dad3ff012468c1251d889a5f5b with parent 43dd3d514b6b11c5195de2fd8e665828801d0972, future [comparison URL](https://perf.rust-lang.org/compare.html?start=43dd3d514b6b11c5195de2fd8e665828801d0972&end=62fd07b1ee7441dad3ff012468c1251d889a5f5b).

rust-timer commented:
Finished benchmarking commit (62fd07b1ee7441dad3ff012468c1251d889a5f5b): [comparison URL](https://perf.rust-lang.org/compare.html?start=43dd3d514b6b11c5195de2fd8e665828801d0972&end=62fd07b1ee7441dad3ff012468c1251d889a5f5b&stat=instructions:u).

### Overall result: ❌ regressions - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=43dd3d514b6b11c5195de2fd8e665828801d0972&end=62fd07b1ee7441dad3ff012468c1251d889a5f5b&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 7.3%     | [0.3%, 63.5%] | 16        |
| Regressions ❌ <br /> (secondary) | 8.7%     | [0.8%, 59.8%] | 8         |
| Improvements ✅ <br /> (primary) | -0.5%    | [-0.5%, -0.5%] | 1         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 6.8%     | [-0.5%, 63.5%] | 17        |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=43dd3d514b6b11c5195de2fd8e665828801d0972&end=62fd07b1ee7441dad3ff012468c1251d889a5f5b&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=43dd3d514b6b11c5195de2fd8e665828801d0972&end=62fd07b1ee7441dad3ff012468c1251d889a5f5b&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|            | mean[^1] | range | count[^2] |
|:----------:|:--------:|:-----:|:---------:|
| Regressions ❌ <br /> (primary) | 15.4%    | [2.2%, 107.2%] | 13        |
| Regressions ❌ <br /> (secondary) | 21.1%    | [2.3%, 110.0%] | 6         |
| Improvements ✅ <br /> (primary) | -        | -     | 0         |
| Improvements ✅ <br /> (secondary) | -        | -     | 0         |
| All ❌✅ (primary) | 15.4%    | [2.2%, 107.2%] | 13        |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

GuillaumeGomez commented:
The impact on performance is just way too huge to be merged as is (+65% on libc!!). We will need to find first what takes so much extra time.

jsha committed:
Generate static file hashes once

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3350960469/jobs/5552099617) [(plain)](https://github.com/rust-lang/rust/commit/a5a61dfa8a7dfbf36b9c55e5c1769555baa0b650/checks/9178371077/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/librustdoc/html/static_files.rs at line 5:
 
 use rustc_data_structures::fx::FxHasher;
 use std::hash::Hasher;
-use std::path::{PathBuf, Path};
+use std::path::{Path, PathBuf};
 use std::{fmt, str};
 pub(crate) struct StaticFile {
 pub(crate) struct StaticFile {
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/librustdoc/html/url_parts_builder.rs" "/checkout/src/librustdoc/html/highlight.rs" "/checkout/src/librustdoc/html/format.rs" "/checkout/src/librustdoc/html/markdown/tests.rs" "/checkout/src/librustdoc/html/static_files.rs" "/checkout/src/librustdoc/html/length_limit.rs" "/checkout/src/librustdoc/html/url_parts_builder/tests.rs" "/checkout/src/librustdoc/html/markdown.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #103714) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 0b0bf10533ce955e6bcc44c2392674f53c2f3952 with merge ed4e3f659f80ce8a5fb718185792c6500325c87c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0b0bf10533ce955e6bcc44c2392674f53c2f3952","merge_sha":"ed4e3f659f80ce8a5fb718185792c6500325c87c"} -->

jsha commented on src/librustdoc/html/render/write_shared.rs:
```diff
@@ -149,139 +44,52 @@ pub(super) fn write_shared(
     }
 
+    // When the user adds their own CSS files with --extend-css, we write that as an
+    // invocation-specific file (that is, with a resource suffix).
```
A couple reasons. One, this is in a technical sense an invocation specific file, since it can be different at each invocation.

Two, from a practical standpoint, it's useful to have the static.files/ directory be exclusively for the toolchain files. For instance, that means that you can run once with `--emit=toolchain-specific`, and know that the contents of static.files/ are complete for that toolchain, and not other content will be served from there.

There's a possible third way, which would be to put a hash in the filename for these CSS files, but serve them from the same directory hierarchy as the invocation-specific files. But I think it's useful to keep things simple and only have two kinds of files (invocation-specific and toolchain).

jsha committed:
Move string literal into format string

Co-authored-by: Michael Howell <michael@notriddle.com>

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3364051432/jobs/5580075202)
Build commit: 6703d43029387eef125596dd7071d0d4fab902c3 (`6703d43029387eef125596dd7071d0d4fab902c3`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3364051432/jobs/5580075202"},"merge_sha":"6703d43029387eef125596dd7071d0d4fab902c3"} -->

rust-timer commented:
Queued 6703d43029387eef125596dd7071d0d4fab902c3 with parent 2afca78a0b03db144c5d8b9f8868feebfe096309, future [comparison URL](https://perf.rust-lang.org/compare.html?start=2afca78a0b03db144c5d8b9f8868feebfe096309&end=6703d43029387eef125596dd7071d0d4fab902c3).

rust-timer commented:
Finished benchmarking commit (6703d43029387eef125596dd7071d0d4fab902c3): [comparison URL](https://perf.rust-lang.org/compare.html?start=2afca78a0b03db144c5d8b9f8868feebfe096309&end=6703d43029387eef125596dd7071d0d4fab902c3&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=2afca78a0b03db144c5d8b9f8868feebfe096309&end=6703d43029387eef125596dd7071d0d4fab902c3&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.6%  | [1.6%, 1.6%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 1.0%  | [1.0%, 1.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-2.7%, -0.3%] | 12    |
| Improvements ✅ <br /> (secondary) | -1.8% | [-2.7%, -0.5%] | 21    |
| All ❌✅ (primary)                 | -0.7% | [-2.7%, 1.6%]  | 13    |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2afca78a0b03db144c5d8b9f8868feebfe096309&end=6703d43029387eef125596dd7071d0d4fab902c3&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.3% | [-2.3%, -2.3%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -2.3% | [-2.3%, -2.3%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=2afca78a0b03db144c5d8b9f8868feebfe096309&end=6703d43029387eef125596dd7071d0d4fab902c3&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1%  | [2.1%, 2.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | 2.1%  | [2.1%, 2.1%]   | 1     |
</details>

notriddle reviewed (approved):
This actually makes things cleaner and faster. I like it!

GuillaumeGomez commented on src/librustdoc/html/layout.rs:
```diff
@@ -34,17 +35,23 @@ pub(crate) struct Page<'a> {
-    pub(crate) fn get_static_root_path(&self) -> &str {
-        self.static_root_path.unwrap_or(self.root_path)
+    pub(crate) fn get_static_root_path(&self) -> String {
+        match self.static_root_path {
```
I just thought about this: instead of recomputing the path every time, wouldn't it be better to instead compute this information even before creating the `Page` struct?

GuillaumeGomez commented:
I added a comment but it's not blocking and it can be done in follow-up PR. In any case, this is a great work, thanks for working on this. r=notriddle,GuillaumeGomez once docs.rs is also ready for this change.

syphar commented:
I just merged & deployed the needed docs.rs change: https://github.com/rust-lang/docs.rs/pull/1889 

cc @notriddle @GuillaumeGomez

jsha commented:
@bors r=notriddle,GuillaumeGomez rollup

bors commented:
:pushpin: Commit c9dbfe31e2e7001571e4d21e46f93d96489df791 has been approved by `notriddle,GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,GuillaumeGomez c9dbfe31e2e7001571e4d21e46f93d96489df791 -->
<!-- homu: {"type":"Approved","sha":"c9dbfe31e2e7001571e4d21e46f93d96489df791","approver":"notriddle,GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
